interface/MO_Interface.h

```diff
@@ -0,0 +1,173 @@
+   * @return Shall returns true if the execute command was accepted. This does not imply successful execution, only acceptance of the command. Shall return false if the execute command can not be accepted
+   *  
+   * TODO: the following section is preliminary and incomplete
+   * In order to report the result of the execute command to the backend, any MO class
+   * implementing this interface must ensure that it will call the DMClient::XXX() method 
+   * exactly once for each time this callback is invoked by the client library and the
+   * callback returned true.
+   *
+   * For each invocation of the execute callback by the client library the implementation
+   * can call DMClient::XXX() either from inside its implementation of the callback
+   * for synchronous reporting, or at *any* time after returning from the callback 
+   * (see section 5.4.1 of the OMA DM 2.0 protocol specification) 
+   * 
+   * TODO: check if additional info (parameters) are needed in order
+   * to correctly link the response (synchronous or asynchronous) with the command.
+   * 
+   * This interface is not considered stable yet
+   */
+  virtual bool execute(const std::string node_path) = 0;
+
+
+  /**
+   *  @}
+   *  @{
+   *  (2) callbacks called by the protocol client library as part of Session or MO management
+   */
+
+  /**
+   * @brief callback for the protocol client library to check MO compatibility
+   *
+   * The protocol client library will call this before adding an instance of an MO implementation
+   * to its MO tree in order to check if the given implementation of the MO class is compatible with
+   * the DDF file registered with the client library. 
+   *
+   * @param[in] ddfname - DDFName as given in the respective name in a DDF file. Example: "urn:oma:mo:oma-dm-dmacc:1.2"
+   * @return - MO implementation classes should only return true if they are compatible with the DDF file corresponding to the given DDFName
+   * 
+   */
+  virtual bool check_ddf_name_compatibility(std::string ddfname) = 0;
+
+  /**
+   * @brief callback on adding the MO instance to the client library's MO tree
+   *
+   * The protocol client library will call this after successfully adding the instance of the MO
+   * into its MO tree.
+   */
+  virtual void init_mo() = 0; // gets called by the client after adding an instance to the tree.
+
+  /**
+   * @brief callback on removing the MO instance from the client library's MO tree
+   *
+   *  Use currently unclear. May be removed later if not needed
+   */
+  virtual void close_mo() = 0; // unclear
+
+  /**
+   * @brief callback to read the miid of the MO instance
+   *
+   * The protocol client library will call this to retrieve the miid of a given MO instance.
+   * The MO classes implementing this interface are responsible for ensuring that miid's are 
+   * unique between all instances of this MO type.
+   */
+  virtual std::string miid() const = 0;
```

Comment:
the MIID should be managed by the client lib. the client app just add MO into the MOtree without maintaining the unique miid.

Reply:
Hmm.... you are right.... It would be more comfortable for MO implementers to handle this in the lib... On the other hand, I want to give the freedom for developers to set their own miid's as well to make things easier understandable when looking at the backend GUI. 
Example: If we have types of firmware on the device that can be updated separately (e.g. roofs, wifi-firmware, camera-firmware), for each I will have an instance of the FUMO MO. Then I will prefer to call the miid "rootfs", "wifi", and "camera", not "1", "2" and "3". So it is good to offer local developer to chose their own firmware.
I will accept your comment and will implement managing of miid in the client lib. I will, however, give an optional parameter for add_MO method so local developers can set their own miid if they want.